### PR TITLE
bakery: add RootKeyStorage

### PR DESCRIPTION
**Pull request by rogpeppe**
This makes it easy to re-use macaroon root keys as desired,
and also to use domain-object-specific root keys to
mint macaroons that are limited in scope.

Even though we re-use public keys, we ensure that all macaroon
ids are unique by appending a nonce to the id.

A subsequent PR will introduce a mongodb-backed implementation of RootKeyStorage.

### CODE REVIEW

File: bakery/service.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ import (
 	"strings"
 
 	"github.com/juju/loggo"
+	"github.com/rogpeppe/fastuuid"
```
Comment (fabricematrat) [inline bakery/service.go:13]: fast because there is a slow one :)

File: bakery/service.go [hunk 2]
```diff
@@ -18,11 +19,14 @@ import (
 
 var logger = loggo.GetLogger("bakery")
 
+var uuidGen = fastuuid.MustNewGenerator()
```
Comment (fabricematrat) [inline bakery/service.go:22]: Does it mean it will panic here ?

File: bakery/storage.go [hunk 3]
```diff
@@ -24,8 +28,71 @@ type Storage interface {
...
+func NewMemRootKeyStorage() RootKeyStorage {
+	return new(memRootKeyStorage)
+}
+
+type memRootKeyStorage struct {
+	mu  sync.Mutex
+	key []byte
+}
+
+func (s *memRootKeyStorage) Get(id string) ([]byte, error) {
```
Comment (fabricematrat) [inline bakery/storage.go:69]: doc ?

File: bakery/storage.go [hunk 4]
```diff
@@ -24,8 +28,71 @@ type Storage interface {
...
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if id != "0" || s.key == nil {
+		return nil, ErrNotFound
+	}
+	return s.key, nil
+}
+
+func (s *memRootKeyStorage) RootKey() (rootKey []byte, id string, err error) {
```
Comment (fabricematrat) [inline bakery/storage.go:80]: doc ?

File: bakery/service.go [hunk 5]
```diff
@@ -213,17 +251,48 @@ func (svc *Service) NewMacaroon(id string, rootKey []byte, caveats []checkers.Ca
...
-	// TODO look at the caveats for expiry time and associate
-	// that with the storage item so that the storage can
-	// garbage collect it at an appropriate time.
-	if err := svc.store.Put(m.Id(), &storageItem{
-		RootKey: rootKey,
-	}); err != nil {
-		return nil, fmt.Errorf("cannot save macaroon to store: %v", err)
+	if svc.rkStore == nil {
+		if err := svc.store.Put(m.Id(), rootKey); err != nil {
+			return nil, fmt.Errorf("cannot save macaroon to store: %v", err)
```
Comment (fabricematrat) [inline bakery/service.go]: I wonder why we use fmt.Errorf as we have a logger declared.

**Comment by fabricematrat** [comment]
:+1:

File: bakery/service.go (hunk 2 above)
Comment (mhilton) [inline bakery/service.go:22]: Only if it can't read from /dev/random. at which point we can't do much anyway.

File: bakery/service.go (hunk 5 above)
Comment (mhilton) [inline bakery/service.go]: fmt.Errorf creates a new error.

File: bakery/storage.go [hunk 6]
```diff
@@ -24,8 +28,71 @@ type Storage interface {
...
+	// Note that the root keys should remain available for as long
+	// as the macaroons using them are valid.
+	//
+	// Note that there is no need for it to return a new root key
+	// for every call - keys may be reused, although some key
+	// cycling is over time is advisable.
+	RootKey() (rootKey []byte, id string, err error)
+}
+
+// NewFallbackStorage returns a RootKeyStorage implementation
```
Comment (mhilton) [inline bakery/storage.go]: This seems to be part of another PR

**Comment by mhilton** [comment]
:+1: with a few thoughts

File: bakery/service.go (hunk 1 above)
Comment (rogpeppe) [inline bakery/service.go:13]: fast 'cos it's a lot faster than just reading from /dev/rand :)

File: bakery/storage.go (hunk 3 above)
Comment (rogpeppe) [inline bakery/storage.go:69]: done

File: bakery/storage.go (hunk 4 above)
Comment (rogpeppe) [inline bakery/storage.go:80]: done

File: bakery/service.go (hunk 5 above)
Comment (rogpeppe) [inline bakery/service.go]: I'll change it to use errgo.Notef.

File: bakery/service.go (hunk 2 above)
Comment (mattyw) [inline bakery/service.go:22]: Even so, macaroon-bakery is just a library, it should be up to the app to decide, but at least it would happen on startup I guess. One concern I have is that depending on the un knowable logging strategy of said app this panic could be logged without any context of where it came from. Why not just do it in NewService?

File: bakery/service.go [hunk 7]
```diff
@@ -40,6 +44,12 @@ type NewServiceParams struct {
 	// an in-memory storage will be used.
 	Store Storage
 
+	// RootKeyStore is used to store macaroon information. If this
```
Comment (mattyw) [inline bakery/service.go]: s/macaroon information/root keys ?

File: bakery/service.go [hunk 8]
```diff
@@ -79,8 +93,27 @@ func NewService(p NewServiceParams) (*Service, error) {
 	return svc, nil
 }
 
+// WithRootKeyStore returns a copy of service where macaroon creation and
+// lookup uses the given store to look up macaroon root keys
```
Comment (mattyw) [inline bakery/service.go]: given root key store. Service has a Store attribute so I'd like to be explicit here

File: bakery/service.go [hunk 9]
```diff
@@ -79,8 +93,27 @@ func NewService(p NewServiceParams) (*Service, error) {
...
+// get root keys and ids.
+//
+// When NewMacaroon is called on the returned Service,
+// is must always be called with an empty id and rootKey.
```
Comment (mattyw) [inline bakery/service.go]: s/is/it

File: bakery/service.go [hunk 10]
```diff
@@ -79,8 +93,27 @@ func NewService(p NewServiceParams) (*Service, error) {
...
+func (svc *Service) WithRootKeyStore(store RootKeyStorage) *Service {
+	svc1 := *svc
+	svc1.rkStore = store
```
Comment (mattyw) [inline bakery/service.go:104]: maybe also

```
svc1.Store = nil
```

**Comment by mattyw** [comment]
:+1: at least one typo, other comment nitpicks and assorted other biksheds

File: bakery/service.go (hunk 7 above)
Comment (rogpeppe) [inline bakery/service.go]: done

File: bakery/service.go (hunk 8 above)
Comment (rogpeppe) [inline bakery/service.go]: done ish

File: bakery/service.go (hunk 9 above)
Comment (rogpeppe) [inline bakery/service.go]: done

File: bakery/service.go (hunk 10 above)
Comment (rogpeppe) [inline bakery/service.go:104]: done
